test/new_relic/healthy_urls_test.rb

```diff
@@ -0,0 +1,138 @@
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+# frozen_string_literal: true
+
+# The unit test class defined herein will verify the health of all URLs found
+# in the project source code.
+#
+# To run the URL health tests by themselves:
+#   TEST=test/new_relic/healthy_urls_test bundle exec rake test
+#
+# A file will be scanned for URLs if the file's basename is found in the
+# FILENAMES array OR the file's extension is found in the EXTENSIONS array
+# unless the file's absolute path matches the IGNORED_FILE_PATTERN regex.
+#
+# NOTE that CHANGELOG.md is handled with special logic so that only the most
+# recent 2 versions mentioned in the changelog are scannned for URLs.
+#
+# See TIMEOUT for the number of seconds permitted for a GET request to a given
+# URL to be completed.
+#
+# Enable DEBUG for additional verbosity
+
+require_relative '../test_helper'
+
+class HealthyUrlsTest < Minitest::Test
+  ROOT = File.expand_path('../../..', __FILE__).freeze
+  FILENAMES = %w[
+    baselines
+    Brewfile
+    Capfile
+    Dockerfile
+    Envfile
+    Gemfile
+    Guardfile
+    install_mysql55
+    LICENSE
+    mega-runner
+    newrelic
+    newrelic_cmd
+    nrdebug
+    Rakefile
+    run_tests
+    runner
+    Thorfile
+  ].freeze
+  EXTENSIONS = %w[
+    css
+    erb
+    gemspec
+    haml
+    html
+    js
+    json
+    md
+    proto
+    rake
+    readme
+    rb
+    sh
+    txt
+    thor
+    tt
+    yml
+  ].freeze
+  FILE_PATTERN = /(?:^(?:#{FILENAMES.join('|')})$)|\.(?:#{EXTENSIONS.join('|')})$/.freeze
+  IGNORED_FILE_PATTERN = %r{/(?:coverage|test)/}.freeze
+  URL_PATTERN = %r{(https?://.*?)[^a-zA-Z0-9/\.\-_#]}.freeze
+  IGNORED_URL_PATTERN = %r{(?:\{|\(|\$|169\.254|\.\.\.|metadata\.google)}
+  TIMEOUT = 5
+  DEBUG = false
+
+  def test_all_urls
+    skip_unless_ci_cron
+    skip_unless_newest_ruby
+    load_httparty
+
+    urls = gather_urls
+    errors = urls.each_with_object({}) do |(url, _files), hash|
+      error = verify_url(url)
+      hash[url] = error if error
+    end
+
+    msg = "#{errors.keys.size} URLs were unreachable!\n\n"
+    msg += errors.map { |url, error| "  #{url} - #{error}\n    files: #{urls[url].join(',')}" }.join("\n")
+
+    assert_empty errors, msg
+  end
+
+  private
+
+  def load_httparty
+    require 'httparty'
+  rescue
+    skip 'Skipping URL health tests in this context, as HTTParty is not available'
+  end
+
+  def real_url?(url)
+    return false if url.match?(IGNORED_URL_PATTERN)
+
+    true
+  end
+
+  def gather_urls
+    Dir.glob(File.join(ROOT, '**', '*')).each_with_object({}) do |file, urls|
```

Comment:
Minor: Can default the `urls` keys to be an empty array, ready to be << onto!

Example from docs: https://rubyapi.org/3.2/o/hash
`h = Hash.new { |hash, key| "Default value for #{key}" }`

Change we can do:
`each_with_object(Hash.new { |h, k| h[k] = [] })`.

Which then would mean that `urls[url] ||= []` could be omitted in  line 155

Reply:
Thanks, @olleolleolle! Implemented with #2132.

Using `Hash.new([])` in conjunction with `+= []` from [Ulysee's blog post](https://medium.com/klaxit-techblog/a-headache-in-ruby-hash-default-values-bf2706660392) looks pretty slick too, but the Ruby docs example approach is more readable and predictable in its behavior, so we'll go with it.

```ruby
LETTERS = ('a'..'z').to_a

hash1 = LETTERS.each_with_object(Hash.new { |h, k| h[k] = [] }) { |l, h| h[l] << true }
hash2 = LETTERS.each_with_object(Hash.new([])) { |l, h| h[l] += [true] }
hash1 == hash2
```